### PR TITLE
Improve media logging support in WandbLogger

### LINKED ISSUE #17940: Include media types other that Image in WandbLogger
### Description & Motivation

Weights & Biases supports [rich media logging](https://docs.wandb.ai/guides/track/log/media) with multiple media types. However, currently the WandbLogger only has [`log_image`](https://github.com/Lightning-AI/lightning/blob/66293758c0192955148e8b4bd9dd156927c23503/src/lightning/pytorch/loggers/wandb.py#L466C9-L466C18) method. This request is to extend the logger to include other media type such as `audio`, and `video`. This makes it easier to log these other media type alongside runs and experiments tracked by the WandbLogger.

### Pitch

Here's an example of how to include audio logging in the logger. 

```python
@rank_zero_only
  def log_audio(self, key: str, audios: List[Any], step: Optional[int] = None, **kwargs: Any) -> None:
      """Log audios (numpy arrays, or file paths).

      Optional kwargs are lists passed to each audio (ex: caption, sample_rate).
      """
      if not isinstance(audios, list):
          raise TypeError(f'Expected a list as "audios", found {type(audios)}')
      n = len(audios)
      for k, v in kwargs.items():
          if len(v) != n:
              raise ValueError(f"Expected {n} items but only found {len(v)} for {k}")
      kwarg_list = [{k: kwargs[k][i] for k in kwargs.keys()} for i in range(n)]
      metrics = {key: [wandb.Audio(audio, **kwarg) for audio, kwarg in zip(audios, kwarg_list)]}
      self.log_metrics(metrics, step)
```

This is very similar to image logging and can be tested as follows in `test_wandb.py`

```python
# test log_audio
  wandb.init().log.reset_mock()
  logger.log_audio(key="samples", audios=["1.mp3", "2.mp3"])
  wandb.Audio.assert_called_with("2.mp3")
  wandb.init().log.assert_called_once_with({"samples": [wandb.Audio(), wandb.Audio()]})

  # test log_audio with step
  wandb.init().log.reset_mock()
  logger.log_audio(key="samples", audios=["1.mp3", "2.mp3"], step=5)
  wandb.Audio.assert_called_with("2.mp3")
  wandb.init().log.assert_called_once_with({"samples": [wandb.Audio(), wandb.Audio()], "trainer/global_step": 5})

  # test log_audio with captions
  wandb.init().log.reset_mock()
  wandb.Audio.reset_mock()
  logger.log_audio(key="samples", audios=["1.mp3", "2.mp3"], caption=["caption 1", "caption 2"])
  wandb.Audio.assert_called_with("2.mp3", caption="caption 2")
  wandb.init().log.assert_called_once_with({"samples": [wandb.Audio(), wandb.Audio()]})

  # test log_audio without a list
  with pytest.raises(TypeError, match="""Expected a list as "audios", found <class 'str'>"""):
      logger.log_audio(key="samples", audios="1.mp3")

  # test log_audio with wrong number of captions
  with pytest.raises(ValueError, match="Expected 2 items but only found 1 for caption"):
      logger.log_audio(key="samples", audios=["1.mp3", "2.mp3"], caption=["caption 1"])
```

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @awaelchli @morganmcg1 @borisdayma @scottire @parambharat

### PR DESCRIPTION
**Pull request by parambharat**
## What does this PR do?

This PR adds two new methods to the `WandbLogger` to log audio and video media types to wandb

Fixes #17940

<details>
  <summary><b>Before submitting</b></summary>

- [X] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/loggers/wandb.py [hunk 1]
```diff
@@ -473,6 +473,38 @@ def log_image(self, key: str, images: List[Any], step: Optional[int] = None, **k
         metrics = {key: [wandb.Image(img, **kwarg) for img, kwarg in zip(images, kwarg_list)]}
         self.log_metrics(metrics, step)  # type: ignore[arg-type]
 
+    @rank_zero_only
+    def log_audio(self, key: str, audios: List[Any], step: Optional[int] = None, **kwargs: Any) -> None:
+        """Log audios (numpy arrays, or file paths).
```
[the remark this answers is not in the document]
Comment (parambharat) [reply src/lightning/pytorch/loggers/wandb.py:482]: Thanks for the review @Borda , I've added the docstrings as advised.

[the remark this answers is not in the document]
Comment (parambharat) [reply src/lightning/pytorch/loggers/wandb.py:491]: Hi, @Borda : Let me know if there is anything I can do to get this merged.

**Comment by parambharat** [comment]
Hi @Borda , just checking in here. Can this be merged ? Let me know if there is anything I can do to help here.

**Comment by parambharat** [comment]
Hi @Borda and @aniketmaurya. Please review and merge this PR. Let me know if there is anything that can be done to help merge this.

**Comment by aniketmaurya** [comment]
hi @Borda, could you please take a look?

**Comment by aniketmaurya** [comment]
hi @parambharat @ayulockin, could you please have a loot at the failing tests.

File: src/lightning/pytorch/loggers/wandb.py [hunk 2]
```diff
@@ -482,6 +484,58 @@ def log_image(self, key: str, images: List[Any], step: Optional[int] = None, **k
...
+
+        Args:
+            key: The key to be used for logging the audio files
+            audios: The list of audio file paths, or numpy arrays to be logged
+            step: The step number to be used for logging the audio files
+            **kwargs: Optional kwargs are lists passed to each Wandb.Audio instance (ex: caption, sample_rate).
```
Comment (carmocca) [inline src/lightning/pytorch/loggers/wandb.py:495]:
```suggestion
            \**kwargs: Optional kwargs are lists passed to each ``Wandb.Audio`` instance (ex: caption, sample_rate).
```

**Review by carmocca — APPROVED** [review]
LGTM

**Comment by parambharat** [comment]
Thanks @aniketmaurya and @carmocca, I've fixed the failing tests introduced by my earlier changes to the logger. However, there are three more tests that are failing.

Here are the details.
```
FAILED loggers/test_wandb.py::test_wandb_pickle - AssertionError
FAILED loggers/test_wandb.py::test_wandb_logger_init_before_spawn - AssertionError
FAILED loggers/test_all.py::test_loggers_fit_test_all[WandbLogger] - AssertionError
```

These tests were modified in the following [commit](https://github.com/Lightning-AI/lightning/commit/b1c83de086ec17ce9e59f4fe979204e816734537) by @awaelchli . I'm not able to pinpoint the reason for the assertion failures. 

@awaelchli can you please help figure out the issue with the tests here ?

File: src/lightning/pytorch/loggers/wandb.py [hunk 3]
```diff
@@ -405,6 +405,8 @@ def experiment(self) -> Union["Run", "RunDisabled"]:
                     self._experiment.define_metric("trainer/global_step")
                     self._experiment.define_metric("*", step_metric="trainer/global_step", step_sync=True)
 
+        assert isinstance(self._experiment, (Run, RunDisabled))
+        self._experiment._label(code="lightning_logger")
```
Comment (awaelchli) [inline src/lightning/pytorch/loggers/wandb.py:409]: Let's avoid calling the protected API here. If this is a feature, we can look into it in a separate PR.

**Review by awaelchli — APPROVED** [review]

**Pull request merged**